petastorm/pytorch.py

```diff
@@ -423,3 +425,160 @@ def __enter__(self):
+        self._shuffling_buffer, other_shuffling_buffer, instantiate_buffer_fn = \
+            self._instantiate_buffers()
+
+        for row in self.reader:
+            # Default collate does not work nicely on namedtuples and treat them as lists
+            # Using dict will result in the yielded structures being dicts as well
+            row_as_dict = row._asdict()
+
+            keys = row_as_dict.keys()
+
+            # Promote some types that are incompatible with pytorch to be pytorch friendly.
+            _sanitize_pytorch_types(row_as_dict)
+
+            # Add rows to shuffling buffer
+            for k, v in row_as_dict.items():
+                if not self.reader.batched_output:
+                    row_as_dict[k] = self.transform_fn([v])
+                else:
+                    row_as_dict[k] = self.transform_fn(v)
+            self._shuffling_buffer.add_many(row_as_dict.values())
+
+            # _yield_batches will emit as much batches as are allowed by the shuffling_buffer (RandomShufflingBuffer
+            # will avoid underflowing below a certain number of samples to guarantee some samples decorrelation)
+            for batch in self._yield_batches(keys):
+                other_shuffling_buffer.add_many(batch.values())
+                yield batch
+
+        # When caching memory, at this point we are done with reading all the data. Rest of the
+        # batches will be created directly from memory.
+        # Otherwise if there is not in memory cache, Once reader can not read new rows, we might
+        # still have a bunch of rows waiting in the shuffling buffer. Telling shuffling buffer
+        # that we are finished allows to deplete the buffer completely, regardless its
+        # min_after_dequeue setting.
+        self._shuffling_buffer.finish()
+
+        for batch in self._yield_batches(keys):
+            other_shuffling_buffer.add_many(batch.values())
+            yield batch
+
+        if self.inmemory_cache_all:
+            for epoch in range(self.num_epochs - 1 if self.num_epochs else sys.maxsize):
+                other_shuffling_buffer.finish()
+                self._shuffling_buffer = other_shuffling_buffer
+                other_shuffling_buffer = instantiate_buffer_fn()
+                for batch in self._yield_batches(keys):
+                    if self.num_epochs is None or self.num_epochs and epoch != self.num_epochs-2:
+                        # We skip populating the other buffer in the last epoch.
+                        other_shuffling_buffer.add_many(batch.values())
+                    yield batch
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        # Close async shuffle thread
+        if self._shuffle_thread:
+            self._close_async_thread()
+        super(AsyncBatchedDataLoader, self).__exit__(exc_type, exc_val, exc_tb)
+
+    def __del__(self):
+        # Close async shuffle thread
+        if self._shuffle_thread:
+            self._close_async_thread()
```

Comment:
If possible, please test-cover this code path?

Reply:
This could be done in a test without using context manager.